### PR TITLE
Destroy fiber in case of exception in cancelBefore in AsyncTaskExecutor

### LINKED ISSUE #55185: RemoteQueryExecutor doesn't look like it's using fibers correctly
EDIT: The first 2/3 of this post are not worth reading, the answer is that fiber destruction calls destructors of local variables by throwing an exception.

RemoteQueryExecutor uses a boost fiber to do async network IO. Unlike threads, suspended fibers can be dropped without calling destructors of local variables. AFAICT (which is not very far tbh), the fiber in RemoteQueryExecutor can be dropped at any time, through this code path:
```
2023.09.29 03:21:51.298510 [ 465038 ] {f66f6914-bb5e-41c9-bc2f-d0514e480b6f} <Error> ExecutingGraph: Code: 209. DB::NetException: Timeout exceeded while connecting to socket (c-cerulean-mw-85-server-0.c-cerulean-mw-85-server-headless.ns-cerulean-mw-85.svc.cluster.local:9000, connection timeout 50 ms). (SOCKET_TIMEOUT), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000ead4ab7 in /usr/bin/clickhouse
1. ? @ 0x0000000013694927 in /usr/bin/clickhouse
2. DB::RemoteQueryExecutorReadContext::checkTimeout(bool) @ 0x000000001369434a in /usr/bin/clickhouse
3. DB::RemoteQueryExecutorReadContext::cancelBefore() @ 0x0000000013694a2d in /usr/bin/clickhouse
4. DB::AsyncTaskExecutor::cancel() @ 0x00000000136933fd in /usr/bin/clickhouse
5. DB::RemoteQueryExecutor::tryCancel(char const*) @ 0x000000001368ca6a in /usr/bin/clickhouse
6. DB::RemoteQueryExecutor::cancel() @ 0x000000001368d080 in /usr/bin/clickhouse
7. DB::ExecutingGraph::cancel(bool) @ 0x00000000156c7d9a in /usr/bin/clickhouse
8. DB::PipelineExecutor::executeStepImpl(unsigned long, std::atomic<bool>*) @ 0x00000000156c2852 in /usr/bin/clickhouse
9. DB::PipelineExecutor::executeImpl(unsigned long) @ 0x00000000156c1d50 in /usr/bin/clickhouse
10. DB::PipelineExecutor::execute(unsigned long) @ 0x00000000156c19e9 in /usr/bin/clickhouse
11. ? @ 0x00000000156c0b4e in /usr/bin/clickhouse
12. ThreadPoolImpl<std::thread>::worker(std::__list_iterator<std::thread, void*>) @ 0x000000000ebbf9af in /usr/bin/clickhouse
13. ? @ 0x000000000ebc5a01 in /usr/bin/clickhouse
14. ? @ 0x00007f138a249b43 in ?
15. ? @ 0x00007f138a2dba00 in ?
 (version 23.8.1.41588 (official build))
```

If that's intentional then all code that executes in that fiber (`RemoteQueryExecutorReadContext::Task::run()`) needs to be written very carefully to avoid doing heap allocations or holding mutexes while suspended. Otherwise the memory and the locks will leak if the fiber gets cancelled while holding those resources. But RemoteQueryExecutorReadContext doesn't seem to be written that way at all. It allocates memory. It locks `cancel_mutex` in `MultiplexedConnections` (which caused a deadlock in prod, which is how I got here).

What's the intended way for this to work?
 * Are all AsyncTaskExecutor users supposed to run their fibers to completion, and it's a bug in RemoteQueryExecutor that it doesn't? This seems unlikely given that `AsyncTaskExecutor::cancel()` and `restart()` exist.
 * Are all RemoteQueryExecutor/ExecutingGraph users supposed to run it to completion, and it's a bug in ExecutingGraph/PipelineExecutor (see stack above) that they don't? This seems unlikely given that `ExecutingGraph::cancel()` and `RemoteQueryExecutor::cancel()` exist.
 * Is RemoteQueryExecutorReadContext supposed to release all resources (like memory or mutexes) on cancellation? But that seems impractically bug-prone, e.g. any temporary local std::string would leak a heap-allocated buffer.
 * Maybe fiber destruction actually does call destructors of local variables? (How??)
 * (More likely I misunderstood everything, and this list makes no sense.)

So, I'm confused. There must be some logic behind it that I didn't understand.

The actual incident that led to this is a "deadlock" where one thread gets stuck here:
```
1. ? @ 0x0000000009eab5c0 in /usr/bin/clickhouse
2. DB::MultiplexedConnections::disconnect() @ 0x00000000155b0f2f in /usr/bin/clickhouse
3. DB::RemoteQueryExecutor::~RemoteQueryExecutor() @ 0x00000000136848e0 in /usr/bin/clickhouse
4. ? @ 0x000000000ecde541 in /usr/bin/clickhouse
5. ? @ 0x0000000013669bc1 in /usr/bin/clickhouse
6. DB::QueryPipeline::~QueryPipeline() @ 0x000000001366c6fa in /usr/bin/clickhouse
7. DB::QueryPipeline::reset() @ 0x000000001367213b in /usr/bin/clickhouse
8. DB::TCPHandler::runImpl() @ 0x00000000156673aa in /usr/bin/clickhouse
9. DB::TCPHandler::run() @ 0x000000001567ccb9 in /usr/bin/clickhouse
10. ? @ 0x000000000eba4f5c in /usr/bin/clickhouse
11. Poco::Net::TCPServerConnection::start() @ 0x0000000018675c54 in /usr/bin/clickhouse
12. Poco::Net::TCPServerDispatcher::run() @ 0x0000000018676e71 in /usr/bin/clickhouse
13. Poco::PooledThread::run() @ 0x00000000187fee67 in /usr/bin/clickhouse
14. Poco::ThreadImpl::runnableEntry(void*) @ 0x00000000187fc85c in /usr/bin/clickhouse
15. ? @ 0x00007f138a249b43 in ?
16. ? @ 0x00007f138a2dba00 in ?
 (version 23.8.1.41588 (official build))
```
after another thread logged the exception above. It gets stuck locking `cancel_mutex`, which is held by the thread that logged the exception. No other thread has MultiplexedConnections in the stack trace, which strongly suggests fiber shenanigans.

Someone please either fix it or explain how it's intended to work so I can fix it.

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix possible deadlock caused by not destroyed fiber in case of exception in async task cancellation. Closes https://github.com/ClickHouse/ClickHouse/issues/55185

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit c0d4b235f74bdffe5a8e508570ccfaf3bae532bc with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55516/c0d4b235f74bdffe5a8e508570ccfaf3bae532bc/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6484975482">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/55516/c0d4b235f74bdffe5a8e508570ccfaf3bae532bc/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by Avogar** [comment]
I could not create a reproducable example unfortunately

**Review by al13n321 — APPROVED** [review]
Lgtm. CC @KochetovNicolai

File: src/Common/AsyncTaskExecutor.cpp [hunk 1]
```diff
@@ -46,7 +46,15 @@ void AsyncTaskExecutor::cancel()
 {
     std::lock_guard guard(fiber_lock);
     is_cancelled = true;
-    cancelBefore();
+    try
```
Comment (alexey-milovidov) [inline src/Common/AsyncTaskExecutor.cpp:49]: You can also use a code block and SCOPE_EXIT.

Comment (alexey-milovidov) [reply src/Common/AsyncTaskExecutor.cpp]: I've edited it, hopefully correctly.

**Review by al13n321 — APPROVED** [review]

**Pull request merged**